core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```
@@ -187,7 +187,7 @@ class TransactionCoordinator(brokerId: Int,
     } else {
       // caller should have synchronized on txnMetadata already
       txnMetadata.state match {
-        case PrepareAbort | PrepareCommit =>
+        case PrepareAbort | PrepareCommit | Dead =>
```

I think we should never see `Dead` as it is only set in the pending state but never in the state field.

---

Ditto below.

---

We should ensure that the `Dead` state is covered by the match. If we are sure it should not happen, then we can raise an exception.

---

The point is that, actually we never "complete the transition to Dead" in the callback, but just remove it from the cache; so `Dead` should only be seen on the pending state field, but never be seen on the state field.

---

related discussions on the previous PR: https://github.com/apache/kafka/pull/3101#discussion_r118397818

---

This is a valid point. In the `TransactionMetadata.completeTransitionTo` method, I now throw an ISE if you try to complete a transition to `Dead`. I guess we can throw the same exception here, but then the client will time out and try again.